modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cluster/ClientClusterGroup.cs

```diff
@@ -0,0 +1,311 @@
+        }
+
+        /** <inheritDoc /> */
+        public ICollection<IClusterNode> GetNodes()
+        {
+            return RefreshNodes();
+        }
+
+        /** <inheritDoc /> */
+        public IClusterNode GetNode(Guid id)
+        {
+            if (id == Guid.Empty)
+            {
+                throw new ArgumentException("Node id should not be empty.");
+            }
+
+            return GetNodes().FirstOrDefault(node => node.Id == id);
+        }
+
+        /** <inheritDoc /> */
+        public IClusterNode GetNode()
+        {
+            return GetNodes().FirstOrDefault();
+        }
+
+        /// <summary>
+        /// Refresh projection nodes.
+        /// </summary>
+        /// <returns>Nodes.</returns>
+        private IList<IClusterNode> RefreshNodes()
+        {
+            long oldTopVer = Interlocked.Read(ref _topVer);
+
+            var topology = RequestTopologyInformation(oldTopVer);
+            if (topology != null)
+            {
+                UpdateTopology(topology.Item1, topology.Item2);
+                UpdateNodesInfo(topology.Item2);
+            }
+
+            // No topology changes.
+            Debug.Assert(_nodeIds != null, "At least one topology update should have occurred.");
+
+            // Local lookup with a native predicate is a trade off between complexity and consistency.
+            var nodes = _nodeIds.Select(id => _ignite.GetNode(id)).Cast<IClusterNode>();
+            if (_predicate != null)
+            {
+                nodes = nodes.Where(_predicate);
+            }
+
+            return nodes.ToList();
+        }
+
+        /// <summary>
+        /// Request topology information.
+        /// </summary>
+        /// <returns>Topology version with nodes identifiers.</returns>rns>
+        private Tuple<long, List<Guid>> RequestTopologyInformation(long oldTopVer)
+        {
+            Action<IBinaryRawWriter> writeAction = writer =>
+            {
+                writer.WriteLong(oldTopVer);
+                _projection.Write(writer);
+            };
+
+            Func<IBinaryRawReader, Tuple<long, List<Guid>>> readFunc = reader =>
+            {
+                if (!reader.ReadBoolean())
+                {
+                    // No topology changes.
+                    return null;
+                }
+
+                long remoteTopVer = reader.ReadLong();
+
+                List<Guid> nodeIds = reader.ReadGuidArray().Cast<Guid>().ToList();
+                return Tuple.Create(remoteTopVer, nodeIds);
+            };
+
+            return DoOutInOp(ClientOp.ClusterGroupGetNodeIds, writeAction, readFunc);
+        }
+
+        /// <summary>
+        /// Update topology.
+        /// </summary>
+        /// <param name="remoteTopVer">Remote topology version.</param>
+        /// <param name="nodeIds">Node ids.</param>
+        internal void UpdateTopology(long remoteTopVer, List<Guid> nodeIds)
+        {
+            lock (this)
```

Comment:
Don't `lock (this)` on objects that are exposed on public API - someone may lock on this object too.
`lock (this)` is an antipattern in general; we use it carefully in some places to reduce allocations, but here we can't do this.